### PR TITLE
Fix CA2016 bugs - diagnostic triggered in edge cases

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/3786

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 1]
```diff
@@ -1881,61 +1934,6 @@ class O
             return CS8VerifyCodeFixAsync(originalCode, fixedCode);
         }
 
-        [Fact]
-        public Task CS_Diagnostic_LambdaAndExtensionMethod()
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:1885]: This test method was converted into a "NoDiagnostic" method.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 2]
```diff
@@ -2002,63 +2000,94 @@ async void M(CancellationToken ct)
             return VerifyCS.VerifyCodeFixAsync(originalCode, fixedCode);
         }
 
-        [Theory]
-        [InlineData("CancellationToken c", "", "", "")]
-        [InlineData("", "CancellationToken c", "", "")]
-        [InlineData("", "", "<CancellationToken>", "c")]
-        public Task CS_Diagnostic_MultiNesting(string topMethodParam, string localMethodParam, string extensionTypeParam, string extensionArg)
+        [Fact]
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:2003]: I split the "MultiNesting" methods into two separate test methods:
- One that verifies the ct being passed from the parent local method.
- One that verifies the ct being passed from the top method, skipping the local method.

The case for the lambda is no longer offering a diagnostic.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 3]
```diff
@@ -3988,31 +3994,22 @@ End Class
             return VerifyVB.VerifyCodeFixAsync(originalCode, fixedCode);
         }
 
-        [Theory]
-        [InlineData("c As CancellationToken", "", "")]
-        [InlineData("", "(Of CancellationToken)", "c")]
-        public Task VB_Diagnostic_MultiNesting(string topMethodParam, string extensionTypeParam, string extensionArg)
+        [Fact]
+        public Task VB_Diagnostic_MultiNesting_TopMethod()
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:3998]: Local methods are not available in VB, but the "MultiNesting" method was still relevant because lock block bodies exist there, so we need to look for the parent method ct.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 4]
```diff
@@ -357,6 +357,59 @@ public static class Extensions
             return CS8VerifyAnalyzerAsync(originalCode);
         }
 
+        [Fact]
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:360]: Nit: Add `WorkItem` attribute for all added tests?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 5]
```diff
@@ -2002,63 +2000,94 @@ async void M(CancellationToken ct)
...
             bool b = false;
-            b.Extension(({extensionArg}) =>
+            lock (lockingObject)
             {{
-                lock (lockingObject)
-                {{
-                    [|TokenMethod|]();
-                }}
-            }});
+                TokenMethod(c);
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:2041]: Do we have a test where this is a static local function? I presume your code fix should bail out for that, and we only get a diagnostic?

Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:2041]: That's a very good edge case.
Let me check. If I don't find such test, I'll add it.
I agree, it should bail out. If it doesn't, I'll fix it.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 6]
```diff
@@ -357,6 +357,59 @@ public static class Extensions
...
+    public static void MyMethod(this int i, CancellationToken c = default)
+    {
+    }
+}
+class C
+{
+    public void M(CancellationToken ct)
+    {
+        bool b = false;
+        b.Extension((j) =>
```
Comment (AArnott) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:403]: Can you add an anonymous delegate version of this test? lambdas and delegates should almost always be treated equivalently, and I think this case is one of those.

Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:403]: Done. For both C# and VB.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs (hunk 4 above)
Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:362]: Done

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs (hunk 5 above)
Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:2092]: I didn't have a test for that, so I added it. The analyzer wasn't able to handle this case, so I fixed it.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 7]
```diff
@@ -357,6 +359,107 @@ public static class Extensions
...
+using System.Threading;
+class C
+{
+    public static void MyMethod(int i, CancellationToken c = default) {}
+    public void M(CancellationToken ct)
+    {
+        LocalStaticMethod();
+        static void LocalStaticMethod()
+        {
+            MyMethod(5);
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: This should generate a diagnostic. We should not be excluding these.

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: Please also add a test where immediate outer local function is not static but its container is a static local function. We should generate a diagnostic even for that case:

```csharp
public static void MyMethod(int i, CancellationToken c = default) {}
    public void M(CancellationToken ct)
    {
        LocalStaticMethod();
        static void LocalStaticMethod()
        {
            LocalMethod();
            void LocalMethod()
            {
                MyMethod(5); // This should generate a diagnostic, but no code fix
            }
        }
}
```

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: Basically, refactoring a large method to have local functions (nested or non-nested ones, static or non-static ones) should not alter the diagnostics reported by this analyzer. Local functions are primarily used to refactor code for readability, unlike lambdas which are passed to other methods and need to ensure no captures to avoid allocations.

Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: @mavasani Can you explain why having a static local function should trigger a diagnostic? According to the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/static-local-functions), the token would not not reachable inside the static local method:

> _"A local function declared static cannot capture state from the enclosing scope. As a result, locals, parameters, and this from the enclosing scope are not available within a static local function."._

I tested the above code in a console app, and I get errors if I attempt to consume the token in any of the two cases:

image

My understanding was that suggesting a diagnostic without offering a fix was for cases that were valid but the fixer had limitations preventing it from offering a fix. In this case, we would be offering a diagnostic for a case that is unsupported/invalid.

Can you please tell me if I'm misunderstanding?

I can add the additional unit test you requested, but I would like to clarify this first.

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: > the token would not not reachable inside the static local method:

Yes, I don't expect a code fix for this case. User would need to either convert it to a non-static local function OR thread in a new cancellation token parameter to the static local function. Both of these are absolutely acceptable changes given it leads to correctly passing in a cancellation token to a cancellable API invoked from the static local method.

Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:456]: I see. Thanks for the explanation.
I will update the analyzer so it offers a diagnostic, and will work on ensuring the fixer bails out when it detects this case.

Comment (carlossanlop) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs]: Addressed, @mavasani. PTAL whenever you get a chance.
